# ã‚¹ãƒ†ãƒ¼ãƒˆãƒã‚·ãƒ³ãƒ»ãƒ“ãƒ˜ã‚¤ãƒ“ã‚¢ï¼ˆStatemachine Behaviourï¼›çŠ¶æ…‹é·ç§»æ©Ÿæ¢°ã®æŒ¯ã‚‹èˆã„ï¼‰
extends Node


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒãƒ¼ãƒ‰ãƒ»ãƒ‘ã‚¹é–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# çŒ¿å–å¾—
func monkey():
	return $"../ğŸµMonkey"


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# èµ·å‹•æ™‚è¨­å®š
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

func _ready():
	self.monkey().statemachine_node().behaviour_of_entry = on_entry
	self.monkey().statemachine_node().behaviour_of_exit = on_exit
	self.monkey().statemachine_node().behaviour_of_go_input = on_go_input
	self.monkey().statemachine_node().behaviour_of_try_inputting_again = on_try_inputting_again
	self.monkey().statemachine_node().behaviour_of_input_accepted = on_input_accepted


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ã‚¤ãƒ™ãƒ³ãƒˆãƒ»ãƒãƒ³ãƒ‰ãƒ©ãƒ¼
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# é–‹å§‹æ™‚ã®æŒ¯ã‚‹èˆã„
func on_entry():
	# ãƒ­ãƒ¼ã‚«ãƒ«ã®ãƒ—ãƒ­ã‚°ãƒ©ãƒãƒ¼ãƒ»ãƒãƒ¼ãƒ‰ã¸ã‚¢ã‚¯ã‚»ã‚¹
	self.monkey().the_programmer_node().on_key_config_entered()


# çµ‚äº†æ™‚ã®æŒ¯ã‚‹èˆã„
func on_exit():
	self.monkey().internal_node().clear_count_by_step()
	
	# ç”»é¢è¡¨ç¤ºã€æ¼”å¥
	self.monkey().display_node().perform_at_close_scene()

	# ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°å‘¼å‡ºã—å…ƒ
	self.monkey().owner_key_config_node().on_exit()


# å…¥åŠ›ã—ã‚
func on_go_input():
	self.monkey().internal_node().clear_count_by_step()


# å…¥åŠ›ã«å†æŒ‘æˆ¦
func on_try_inputting_again(reason):
	# ã‚­ãƒ£ãƒ³ã‚»ãƒ«ãƒ»ãƒœã‚¿ãƒ³æŠ¼ä¸‹å¾Œã€å†å…¥åŠ›
	if reason == &"CancelButtonPushed":
		# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°é …ç›®ã‚’ã€ï¼‘ã¤ã‹ã€ï¼’ã¤ï¼ˆãƒ¬ãƒãƒ¼æ™‚ï¼‰æˆ»ã™
		self.monkey().display_node().show_cancel()
		self.monkey().display_node().set_empty_the_button_message(self.monkey().internal_node().key_config_item_number)

		self.monkey().internal_node().key_config_item_number -= 1
		# ã•ã‚‰ã«é€£ç¶šã—ã¦æˆ»ã—ãŸã„ã‚±ãƒ¼ã‚¹ã‚‚ã‚ã‚‹
		# ãƒ¬ãƒãƒ¼ã®ä¸Š
		if self.monkey().internal_node().key_config_item_number == 5 and self.monkey().owner_key_config_node().key_config[&"VK_Down"] == self.monkey().owner_key_config_node().key_config[&"VK_Up"]:
			self.monkey().display_node().set_empty_the_button_message(self.monkey().internal_node().key_config_item_number)
			self.monkey().internal_node().key_config_item_number -= 1
			self.monkey().owner_key_config_node().key_config.erase(&"VK_Down")
		# ãƒ¬ãƒãƒ¼ã®å·¦
		elif self.monkey().internal_node().key_config_item_number == 7 and self.monkey().owner_key_config_node().key_config[&"VK_Right"] == self.monkey().owner_key_config_node().key_config[&"VK_Left"]:
			self.monkey().display_node().set_empty_the_button_message(self.monkey().internal_node().key_config_item_number)
			self.monkey().internal_node().key_config_item_number -= 1
			self.monkey().owner_key_config_node().key_config.erase(&"VK_Right")
		
		self.monkey().display_node().set_press_message_to_button(self.monkey().internal_node().key_config_item_number)
		
		if self.monkey().internal_node().previous_virtual_key_name != null:
			self.monkey().owner_key_config_node().key_config.erase(self.monkey().internal_node().previous_virtual_key_name)
		
		self.monkey().internal_node().clear_count_by_step()

	# æ—¢å­˜ã®ã‚­ãƒ¼ã¨è¢«ã£ã¦ã€å†å…¥åŠ›
	elif reason == &"KeyDuplicated":
		self.monkey().display_node().show_pushed_button_denied(1)
		self.monkey().internal_node().clear_count_by_step()
	
	# ã‚¤ãƒ³ã‚¿ãƒ¼ãƒãƒ«å¾Œ
	elif reason == &"AfterInterval":
		self.monkey().display_node().set_press_message_to_button(self.monkey().internal_node().key_config_item_number)

	# ä¸‹ã‚­ãƒ¼ãŒãƒœã‚¿ãƒ³ãªã‚‰ã€ä¸Šãƒœã‚¿ãƒ³ã‚‚å†å…¥åŠ›
	elif reason == &"SelectUpButton":
		self.monkey().display_node().show_pushed_button_denied(2)

	# å³ã‚­ãƒ¼ãŒãƒœã‚¿ãƒ³ãªã‚‰ã€å·¦ãƒœã‚¿ãƒ³ã‚‚å†å…¥åŠ›
	elif reason == &"SelectLeftButton":
		self.monkey().display_node().show_pushed_button_denied(3)

	else:
		print("ï¼»ã‚­ãƒ¼ã‚³ãƒ³ãƒ•ã‚£ã‚°ï¼½ã€€â–²ï¼ã€€ã‚¨ãƒ©ãƒ¼ã€€reason:" + reason)


# å…¥åŠ›ã‚’å—ã‘ä»˜ã‘ãŸ
func on_input_accepted():
	# ç™»éŒ²
	self.monkey().owner_key_config_node().push_virtual_key(
			self.monkey().internal_node().virtual_key_name,
			self.monkey().internal_node().button_symbol)
	# æ±ºå®šæ¼”å‡º
	self.monkey().display_node().show_pushed_button_accepted(
			self.monkey().internal_node().key_config_item_number,
			self.monkey().internal_node().button_symbol)

	# ãƒ¬ãƒãƒ¼ã®ä¸‹
	if self.monkey().internal_node().key_config_item_number == 4:
		var hardware_symbol = self.monkey().owner_key_config_node().key_config[&"VK_Down"]
		if typeof(hardware_symbol) != TYPE_STRING && 1000 <= hardware_symbol:
			# è»¸ã‚’é¸æŠã—ãŸãªã‚‰ã€ãƒ¬ãƒãƒ¼ã®ä¸Šã®é¸æŠã¯ã‚¹ã‚­ãƒƒãƒ—
			# ä¸Šã‚­ãƒ¼ã«ã‚‚åŒæ§˜ã«ç™»éŒ²
			self.monkey().owner_key_config_node().push_virtual_key(&"VK_Up", self.monkey().internal_node().button_symbol)
			# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸è¡¨ç¤º
			self.monkey().display_node().set_done_message_the_button(
					self.monkey().internal_node().key_config_item_number + 1,
					self.monkey().internal_node().button_symbol)
			self.monkey().internal_node().key_config_item_number += 2
		else:
			self.monkey().internal_node().key_config_item_number += 1
			
	# ãƒ¬ãƒãƒ¼ã®å³
	elif self.monkey().internal_node().key_config_item_number == 6:
		var hardware_symbol = self.monkey().owner_key_config_node().key_config[&"VK_Right"]
		if typeof(hardware_symbol) != TYPE_STRING && 1000 <= hardware_symbol:
			# è»¸ã‚’é¸æŠã—ãŸãªã‚‰ã€ãƒ¬ãƒãƒ¼ã®å·¦ã®é¸æŠã¯ã‚¹ã‚­ãƒƒãƒ—
			# å·¦ã‚­ãƒ¼ã«ã‚‚åŒæ§˜ã«ç™»éŒ²
			self.monkey().owner_key_config_node().push_virtual_key(&"VK_Left", self.monkey().internal_node().button_symbol)
			# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸è¡¨ç¤º
			self.monkey().display_node().set_done_message_the_button(
					self.monkey().internal_node().key_config_item_number + 1,
					self.monkey().internal_node().button_symbol)
			self.monkey().internal_node().key_config_item_number += 2
		else:
			self.monkey().internal_node().key_config_item_number += 1

	else:
		self.monkey().internal_node().key_config_item_number += 1
